File: Python/05. Inheritance Relationship/practice_18.py
"""
Practice 18: Vehicle Rental — Method Overriding & Constructor Chaining
Task: Enhance the vehicle rental system with method overriding and
      constructor chaining between base class and subclasses.

Changes from Practice 17:
  - Constructor chaining: subclass constructors call super().__init__()
  - Method overriding: calculate_rental_cost() defined in Vehicle base class,
    overridden in each subclass
  - __str__ overriding: each class provides its own string representation

How to run:
  python practice_18.py

Key Concepts:
  - Method overriding
  - Constructor chaining (super().__init__())
  - __str__ / toString overriding
"""


class Vehicle:
    """Base class for all rental vehicles."""

    def __init__(self, brand: str, model: str, year: int):
        self.brand = brand
        self.model = model
        self.year = year

    def calculate_rental_cost(self, days: int) -> float:
        """Base rental cost calculation — overridden by subclasses."""
        return 0.0

    def __str__(self) -> str:
        return f"{self.brand} {self.model} ({self.year})"


class Car(Vehicle):
    """Car: $50/day, 10% discount if older than 5 years."""

    BASE_RATE = 50.0

    def __init__(self, brand: str, model: str, year: int):
        super().__init__(brand, model, year)  # Constructor chaining

    def calculate_rental_cost(self, days: int) -> float:
        """Overrides Vehicle.calculate_rental_cost()."""
        cost = self.BASE_RATE * days
        if 2026 - self.year > 5:
            cost *= 0.9
        return cost

    def __str__(self) -> str:
        return f"[Car] {super().__str__()}"


class Bike(Vehicle):
    """Bike: $15/day, 15% discount if rental > 7 days."""

    BASE_RATE = 15.0

    def __init__(self, brand: str, model: str, year: int):
        super().__init__(brand, model, year)  # Constructor chaining

    def calculate_rental_cost(self, days: int) -> float:
        """Overrides Vehicle.calculate_rental_cost()."""
        cost = self.BASE_RATE * days
        if days > 7:
            cost *= 0.85
        return cost

    def __str__(self) -> str:
        return f"[Bike] {super().__str__()}"


class Truck(Vehicle):
    """Truck: $100/day + $100/day additional fee."""

    BASE_RATE = 100.0
    ADDITIONAL_FEE = 100.0

    def __init__(self, brand: str, model: str, year: int):
        super().__init__(brand, model, year)  # Constructor chaining

    def calculate_rental_cost(self, days: int) -> float:
        """Overrides Vehicle.calculate_rental_cost()."""
        return (self.BASE_RATE + self.ADDITIONAL_FEE) * days

    def __str__(self) -> str:
        return f"[Truck] {super().__str__()}"


def main():
    car = Car("Toyota", "Corolla", 2019)
    bike = Bike("Yamaha", "R15", 2023)
    truck = Truck("Volvo", "FH16", 2022)

    days = 10

    # --- Method overriding: calculate_rental_cost() ---
    print("=== Vehicle Rental — Method Overriding ===")
    print()
    for vehicle in [car, bike, truck]:
        print(f"  Vehicle : {vehicle}")                                         # __str__ override
        print(f"  Rental ({days} days): ${vehicle.calculate_rental_cost(days):.2f}")  # Method override
        print()

    # --- __str__ overriding ---
    print("=== String Representation (__str__) ===")
    print(f"  Car   : {car}")
    print(f"  Bike  : {bike}")
    print(f"  Truck : {truck}")
    print()

    # --- Discount scenarios ---
    print("=== Discount Scenarios ===")

    old_car = Car("Honda", "Civic", 2018)
    print(f"  {old_car}")
    print(f"  Age: {2026 - old_car.year} years → 10% discount applied")
    print(f"  Rental ({days} days): ${old_car.calculate_rental_cost(days):.2f}")
    print()

    short_bike = Bike("Honda", "CBR", 2024)
    short_days = 5
    print(f"  {short_bike}")
    print(f"  Rental ({short_days} days): ${short_bike.calculate_rental_cost(short_days):.2f} (no discount)")
    print()

    long_bike = Bike("Suzuki", "Gixxer", 2023)
    long_days = 10
    print(f"  {long_bike}")
    print(f"  Rental ({long_days} days): ${long_bike.calculate_rental_cost(long_days):.2f} (15% discount)")


if __name__ == "__main__":
    main()

File: Java/05. Inheritance Relationship/Practice18.java
/**
 * Practice 18: Vehicle Rental — Method Overriding & Constructor Chaining
 * Task: Enhance the vehicle rental system with method overriding and
 *       constructor chaining between base class and subclasses.
 *
 * Changes from Practice 17:
 *   - Constructor chaining: subclass constructors call super()
 *   - Method overriding: calculateRentalCost() in Vehicle, overridden in subclasses
 *   - toString() overriding in each class
 *
 * How to compile and run:
 *   javac Practice18.java
 *   java Practice18
 *
 * Key Concepts:
 *   - Method overriding (@Override)
 *   - Constructor chaining (super())
 *   - toString() overriding
 */

public class Practice18 {

    /** Base class for all rental vehicles. */
    static class Vehicle {
        protected String brand;
        protected String model;
        protected int year;

        public Vehicle(String brand, String model, int year) {
            this.brand = brand;
            this.model = model;
            this.year = year;
        }

        public double calculateRentalCost(int days) {
            return 0.0;
        }

        @Override
        public String toString() {
            return brand + " " + model + " (" + year + ")";
        }
    }

    /** Car: $50/day, 10% discount if older than 5 years. */
    static class Car extends Vehicle {
        private static final double BASE_RATE = 50.0;

        public Car(String brand, String model, int year) {
            super(brand, model, year);  // Constructor chaining
        }

        @Override
        public double calculateRentalCost(int days) {
            double cost = BASE_RATE * days;
            if (2026 - year > 5) {
                cost *= 0.9;
            }
            return cost;
        }

        @Override
        public String toString() {
            return "[Car] " + super.toString();
        }
    }

    /** Bike: $15/day, 15% discount if rental > 7 days. */
    static class Bike extends Vehicle {
        private static final double BASE_RATE = 15.0;

        public Bike(String brand, String model, int year) {
            super(brand, model, year);  // Constructor chaining
        }

        @Override
        public double calculateRentalCost(int days) {
            double cost = BASE_RATE * days;
            if (days > 7) {
                cost *= 0.85;
            }
            return cost;
        }

        @Override
        public String toString() {
            return "[Bike] " + super.toString();
        }
    }

    /** Truck: $100/day + $100/day additional fee. */
    static class Truck extends Vehicle {
        private static final double BASE_RATE = 100.0;
        private static final double ADDITIONAL_FEE = 100.0;

        public Truck(String brand, String model, int year) {
            super(brand, model, year);  // Constructor chaining
        }

        @Override
        public double calculateRentalCost(int days) {
            return (BASE_RATE + ADDITIONAL_FEE) * days;
        }

        @Override
        public String toString() {
            return "[Truck] " + super.toString();
        }
    }

    public static void main(String[] args) {
        Vehicle car = new Car("Toyota", "Corolla", 2019);
        Vehicle bike = new Bike("Yamaha", "R15", 2023);
        Vehicle truck = new Truck("Volvo", "FH16", 2022);

        int days = 10;

        System.out.println("=== Vehicle Rental — Method Overriding ===");
        System.out.println();

        for (Vehicle vehicle : new Vehicle[]{car, bike, truck}) {
            System.out.println("  Vehicle : " + vehicle);                                  // toString override
            System.out.printf("  Rental (%d days): $%.2f%n", days, vehicle.calculateRentalCost(days)); // Method override
            System.out.println();
        }

        System.out.println("=== String Representation (toString) ===");
        System.out.println("  Car   : " + car);
        System.out.println("  Bike  : " + bike);
        System.out.println("  Truck : " + truck);
        System.out.println();

        System.out.println("=== Discount Scenarios ===");

        Vehicle oldCar = new Car("Honda", "Civic", 2018);
        System.out.println("  " + oldCar);
        System.out.printf("  Age: %d years → 10%% discount applied%n", 2026 - oldCar.year);
        System.out.printf("  Rental (%d days): $%.2f%n", days, oldCar.calculateRentalCost(days));
        System.out.println();

        Vehicle shortBike = new Bike("Honda", "CBR", 2024);
        int shortDays = 5;
        System.out.println("  " + shortBike);
        System.out.printf("  Rental (%d days): $%.2f (no discount)%n", shortDays, shortBike.calculateRentalCost(shortDays));
        System.out.println();

        Vehicle longBike = new Bike("Suzuki", "Gixxer", 2023);
        int longDays = 10;
        System.out.println("  " + longBike);
        System.out.printf("  Rental (%d days): $%.2f (15%% discount)%n", longDays, longBike.calculateRentalCost(longDays));
    }
}
